Suggest casting on numeric type error

estebank opened the pull request:
Re #47168.

estebank committed:
Suggest casting on numeric type error

rust-highfive commented:
r? @petrochenkov

(rust_highfive has picked a reviewer for you, use r? to override)

GuillaumeGomez commented:
Shouldn't we suggest to use `from` instead of casting?

GuillaumeGomez commented:
More globally, I'm not sure talking about size fit/shortage is a good idea. Just talking about `from` should be enough for most cases, don't you think?

varkor commented on src/librustc_typeck/check/demand.rs:
```diff
@@ -306,12 +304,181 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
+                                            &format!("{}, producing the closest possible value",
+                                                     msg),
+                                            suggestion);
+                        err.warn("casting here will cause Undefined Behavior if the value is \
```
Should "undefined behaviour" be capitalised like this? [It doesn't seem to be](https://github.com/rust-lang/rust/blob/565907fefb2a69c279b59bffa84c6d30129cafe9/src/librustc_lint/builtin.rs#L1236) typical in error messages.

estebank commented:
> More globally, I'm not sure talking about size fit/shortage is a good idea.

I find the possibility of using the compiler errors to teach the language appealing. Presenting some of the possible trade offs inline like this seems like a great opportunity, even though completely I agree with the problems presented by `as`.

> Just talking about from should be enough for most cases, don't you think?

Part of what gave me pause is that `From` is not implemented for all conversions, only for the ones that cannot fail, and `TryFrom` is confined to nightly behind a feature flag. Also, the suggestion becomes much more verbose in that case (and only applies to nightly):

```
fn foo<T>(x: T) {}
fn main() {
    foo::<u32>(3_i32);
}
```
after:
```
#![feature(try_from)]
use std::convert::TryFrom;

fn foo<T>(x: T) {}
fn main() {
    foo::<u32>(u32::try_from(3_i32).unwrap());
}
```

That being said, the way I wrote the code was to make it possible to only show a subset of these suggestions depending on the case. I could suggest `into()` when it is supported (growing an type, `u16` to `u32`), and remain silent in the other cases.

estebank committed:
Use `into` for casting when possible

estebank committed:
Only suggest casting numeric types using `into()`

estebank commented:
Updated the PR to only suggest `.into()` conversions. The code for suggesting casts is still there, just never triggered (it is gated on the type of expression, right now not enabled for any).

GuillaumeGomez commented:
That's problematic that `TryFrom` is still nightly only... And where is `From` implementation missing?

estebank commented:
@GuillaumeGomez in https://github.com/rust-lang/rust/pull/47247/files#diff-1743dc5c92f33d15fb1e903d7ad3ce58 (which is admittedly very verbose, as it has every single possible numeric type mismatch) you can see that the only suggestions provided are for the types that have a `From` implementation:

 - within the same numeric type, being signed or unsigned, there's a `From` implementation from the type with a lesser bit width to the wider type (`u16` -> `u32`, `i32` -> `i64`),
 - from `f32` to `f64`,
 - from 32 bit integers and lower to `f64`
 - from 16 bit integers and lower to `f32`.

petrochenkov commented on src/librustc_typeck/check/demand.rs:
```diff
@@ -306,12 +304,251 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
+        let can_cast = false;
+
+        let needs_paren = match expr.node {
+            hir::ExprBinary(..) => true,
```
I think you can reuse `syntax::util::parser::expr_precedence` here and get more precise result.
See `fn print_expr_method_call` in `pprust.rs` for an example.

estebank replied:
Updated.

petrochenkov commented on src/librustc_typeck/check/demand.rs:
```diff
@@ -306,12 +304,251 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
+                                            cast_suggestion);
+                        err.warn("casting here will cause undefined behavior if the value is \
+                                  finite but larger or smaller than the largest or smallest \
+                                  finite value representable by `f32` (this is a bug and will be \
```
float <-> float conversions are buggy?
I don't think this is true. There's [`-Z saturating-float-casts`](https://github.com/rust-lang/rust/pull/45205) for correcting float -> int casts, but everything else is already correct, if I remember correctly.

estebank replied:
This warning is taken from https://doc.rust-lang.org/book/first-edition/casting-between-types.html#numeric-casts. Is it out of date?

petrochenkov replied:
Oh, indeed, https://github.com/rust-lang/rust/issues/15536 is still open.
Ping @rkruppe just in case.

rkruppe replied:
The default behavior is indeed still UB, saturating casts are still opt-in.

petrochenkov commented on src/librustc_typeck/check/demand.rs:
```diff
@@ -306,12 +304,251 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
+                    if exp.bit_width() > found.bit_width().unwrap_or(256) {
+                        err.span_suggestion(expr.span,
+                                            &format!("{}, producing the floating point \
+                                                      representation of the integer, rounded if \
```
If `into` is available, then the conversion will be lossless and no rounding will happen.
Also `into` is not available for `i/usize` and floats.

estebank replied:
Updated message.

The suggestion is not given for `isize`/`usize` (that's what the `unwrap(256)` is doing). Added comment to that effect.

petrochenkov commented on src/librustc_typeck/check/demand.rs:
```diff
@@ -306,12 +304,251 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
+                    if exp.bit_width() > found.bit_width().unwrap_or(256) {
+                        err.span_suggestion(expr.span,
+                                            &format!("{}, producing the floating point \
+                                                      representation of the integer, rounded if \
```
Same as https://github.com/rust-lang/rust/pull/47247/files#r160534951

estebank replied:
Done.

petrochenkov commented:
Looks pretty useful.
I'm also okay with dead code from `can_cast = false` because the implementation doesn't look "complete" without it, something can be done with it later.
I left some comments inline.

estebank committed:
Use `s::u::p::expr_precedence` and fix message

 - Use `syntax::util::parser::expr_precedence` to determine wether
   parenthesis are needed around the casting target.
 - Update message to not incorrectly mention rounding on `.into()`
   suggestions, as those types that do have that implemented will never
   round.

petrochenkov commented on src/librustc_typeck/check/demand.rs:
```diff
@@ -306,12 +305,248 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
+        // For now, don't suggest casting with `as`.
+        let can_cast = false;
+
+        let needs_paren = expr_precedence(expr) < (AssocOp::As.precedence() as i8);
```
Argh, this `expr_precedence` works on AST.
There's another one in `rustc::hir::print`.
Also, `AssocOp::As.precedence()` -> `PREC_POSTFIX`.

estebank replied:
I'm gonna add a commit to merge both into one place (new enum that only holds the type to precedence mapping, and an into impl for `{hir, ast}::Expr` so that there's only one source of truth for this. Started that yesterday but went to bed before I could complete it.

estebank committed:
Use single source of truth for expr precedence

Introduce a new unified type that holds the expression precedence for
both AST and HIR nodes.

bors commented:
:umbrella: The latest upstream changes (presumably #46461) made this pull request unmergeable. Please resolve the merge conflicts.

estebank commented:
Fixed.

petrochenkov commented on src/libsyntax/ast.rs:
```diff
@@ -903,6 +917,129 @@ pub struct Expr {
 }
 
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum ExprPrecedence {
```
Could you move all this stuff from `ast.rs` somewhere, e.g. into `util/parser.rs` where it was previously.
This is not a part of AST or HIR, really, it's a parsing-specific detail.

estebank replied:
Done. Waiting for the tests to run to make sure nothing broke.

estebank committed:
Move `ExprPrecedence` to `libsyntax/util/parser.rs`

petrochenkov commented:
@bors r+

bors commented:
:pushpin: Commit 71c0873 has been approved by `petrochenkov`

<!-- @bors r=petrochenkov 71c08734a3ef12135ba695964e6fe882bc5e6d6e -->
